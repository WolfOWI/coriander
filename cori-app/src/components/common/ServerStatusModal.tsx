import React, { useEffect, useState } from "react";
import { Modal, Spin } from "antd";
import { healthCheckAPI } from "../../services/api.service";

interface ServerStatusModalProps {
  isVisible: boolean;
  onClose: () => void;
}

const ServerStatusModal: React.FC<ServerStatusModalProps> = ({ isVisible, onClose }) => {
  const [isChecking, setIsChecking] = useState(true);
  const [isServerAwake, setIsServerAwake] = useState(false);
  const [checkCount, setCheckCount] = useState(0);

  // Function to check if server is awake using lightweight health check
  const checkServerStatus = async () => {
    console.log("🔍 Modal: Checking server status...");
    setIsChecking(true);
    setCheckCount((prev) => prev + 1);
    try {
      await healthCheckAPI.checkHealth();
      console.log("✅ Modal: Server is awake");
      setIsServerAwake(true);
      // Add a small delay to show the "Server is back online" state
      await new Promise((resolve) => setTimeout(resolve, 1000));
      console.log("⏰ Showing 'Server is back online' state");
      // Then wait another 2 seconds before closing
      setTimeout(() => {
        console.log("⏰ Delay complete, closing modal");
        setIsChecking(false);
        onClose();
      }, 2000);
    } catch (error) {
      console.log("💤 Modal: Server is sleeping");
      // Add a small delay before showing sleeping state
      await new Promise((resolve) => setTimeout(resolve, 3000));
      setIsServerAwake(false);
      setIsChecking(false);
    }
  };

  // When modal becomes visible, start checking server status
  useEffect(() => {
    let interval: NodeJS.Timeout;
    if (isVisible) {
      console.log("👁️ Modal became visible, starting checks");
      setCheckCount(0); // Reset check count when modal becomes visible
      checkServerStatus();
      // Check every 10 seconds while modal is visible
      interval = setInterval(checkServerStatus, 10000);
    }
    return () => {
      if (interval) {
        console.log("👋 Modal cleanup - clearing interval");
        clearInterval(interval);
      }
    };
  }, [isVisible]);

  return (
    <Modal
      open={isVisible}
      onCancel={onClose}
      footer={null}
      closable={!isChecking}
      maskClosable={!isChecking}
    >
      <div className="flex flex-col items-center justify-center p-6 gap-4 text-center">
        <h1 className="text-4xl font-bold">Oops!</h1>
        <h3 className="text-lg">The server is not responding. It is either offline or sleeping.</h3>
        <Spin size="large" spinning={isChecking} />
        <h3 className="text-lg font-semibold">
          {isChecking
            ? checkCount === 1
              ? "Checking Server Status..."
              : `Let's Try Again (Attempt ${checkCount})...`
            : isServerAwake
            ? "Server is back online!"
            : "Server is currently sleeping"}
        </h3>
        <p className="text-gray-600 text-center">
          {!isServerAwake &&
            !isChecking &&
            "The server is waking up. This may take up to 50 seconds..."}
        </p>
      </div>
    </Modal>
  );
};

export default ServerStatusModal;
