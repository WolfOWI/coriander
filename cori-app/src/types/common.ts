/**
 * Gender enum matching the backend values
 */
export enum Gender {
  Male = 0,
  Female = 1,
  Other = 2,
}

/**
 * Pay Cycle enum matching the backend values
 */
export enum PayCycle {
  Monthly = 0,
  BiWeekly = 1,
  Weekly = 2,
}

/**
 * Employ Type enum matching the backend values
 */
export enum EmployType {
  FullTime = 0,
  PartTime = 1,
  Contract = 2,
  Intern = 3,
}

/**
 * Equipment Condition enum matching the backend values
 */
export enum EquipmentCondition {
  New = 0,
  Good = 1,
  Decent = 2,
  Used = 3,
}

/**
 * Equipment Category enum matching the backend values
 */
export enum EquipmentCategory {
  Cellphone = 1,
  Tablet = 2,
  Laptop = 3,
  Monitor = 4,
  Headset = 5,
  Keyboard = 6,
}
/**
 * Review Status enum matching the backend values
 */
export enum ReviewStatus {
  Pending = 0,
  Upcoming = 1,
  Completed = 2,
}

/**
 * User Role enum matching the backend values
 */
export enum UserRole {
  Unassigned = 0,
  Employee = 1,
  Admin = 2,
}

export enum LeaveStatus {
  Pending = 0,
  Approved = 1,
  Rejected = 2,
}
