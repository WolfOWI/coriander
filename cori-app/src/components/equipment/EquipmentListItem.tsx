import React, { JSX } from "react";

// Import Components
import CoriBadge from "../badges/CoriBadge";
import CoriCircleBtn from "../buttons/CoriCircleBtn";

// Functions
import dayjs from "dayjs";

// Import Icons
import { Icons } from "../../constants/icons";

// Types
import { EquipmentCondition, EquipmentCategory } from "../../types/common";
import EquipCondiBadge from "../badges/EquipCondiBadge";

interface Equipment {
  equipmentId: number;
  employeeId: number;
  equipmentCatId: EquipmentCategory;
  equipmentCategoryName: string;
  equipmentName: string;
  assignedDate: string;
  condition: EquipmentCondition;
}

interface EquipmentListItemProps {
  item: Equipment | null;
  onEdit?: () => void;
  adminView?: boolean;
}

function EquipmentListItem({ item, onEdit, adminView }: EquipmentListItemProps) {
  let deviceIcon: JSX.Element;

  if (!item) {
    return null;
  }

  switch (item.equipmentCatId) {
    case EquipmentCategory.Cellphone:
      deviceIcon = <Icons.Phone fontSize="large" />;
      break;
    case EquipmentCategory.Tablet:
      deviceIcon = <Icons.Tablet fontSize="large" />;
      break;
    case EquipmentCategory.Laptop:
      deviceIcon = <Icons.Laptop fontSize="large" />;
      break;
    case EquipmentCategory.Monitor:
      deviceIcon = <Icons.Monitor fontSize="large" />;
      break;
    case EquipmentCategory.Headset:
      deviceIcon = <Icons.Headset fontSize="large" />;
      break;
    case EquipmentCategory.Keyboard:
      deviceIcon = <Icons.Keyboard fontSize="large" />;
      break;
    default:
      deviceIcon = <Icons.DeviceUnknown fontSize="large" />;
      break;
  }

  return (
    <div className="flex items-center gap-2 w-full justify-between group">
      <div className="flex items-center gap-2">
        {deviceIcon}
        <div className="flex flex-col">
          <p className="text-zinc-900">{item.equipmentName}</p>
          <p className="text-zinc-500 text-sm">{item.equipmentCategoryName}</p>
        </div>
      </div>
      <p className="text-zinc-500 text-sm">{dayjs(item.assignedDate).format("DD MMM YYYY")}</p>
      <div className="flex items-center gap-2">
        <EquipCondiBadge condition={item.condition} />
        {/* Edit button only for admins */}
        {adminView && (
          <>
            <CoriCircleBtn
              style="black"
              icon={<Icons.Edit />}
              onClick={onEdit}
              className="hidden group-hover:flex transition-all duration-300"
            />
          </>
        )}
      </div>
    </div>
  );
}

export default EquipmentListItem;
