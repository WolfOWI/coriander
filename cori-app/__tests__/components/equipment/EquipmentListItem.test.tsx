import React from "react";
import "@testing-library/jest-dom";
import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import EquipmentListItem from "../../../src/components/equipment/EquipmentListItem";
import { EquipmentCategory, EquipmentCondition } from "../../../src/types/common";

const mockEquipment = {
  equipmentId: 1,
  employeeId: null,
  equipmentCatId: EquipmentCategory.Laptop,
  equipmentCategoryName: "Laptop",
  equipmentName: "MacBook Pro",
  assignedDate: "2024-03-20",
  condition: EquipmentCondition.Good,
};

describe("Component Tests for EquipmentListItem", () => {
  test("EquipmentListItem should render correctly with basic props", () => {
    render(<EquipmentListItem item={mockEquipment} />);

    // Check if basic information is displayed
    expect(screen.getByText(mockEquipment.equipmentName)).toBeInTheDocument();
    expect(screen.getByText(mockEquipment.equipmentCategoryName)).toBeInTheDocument();
    expect(screen.getByText("20 Mar 2024")).toBeInTheDocument();
  });

  test("EquipmentListItem should render admin buttons when adminView is true", async () => {
    const mockOnEdit = jest.fn();
    const mockOnUnlink = jest.fn();
    const mockOnDelete = jest.fn();
    const user = userEvent.setup();

    render(
      <EquipmentListItem
        item={mockEquipment}
        adminView={true}
        onEdit={mockOnEdit}
        onUnlink={mockOnUnlink}
        onDelete={mockOnDelete}
      />
    );

    // The buttons should be in the document but initially hidden
    const editButton = screen.getByRole("button", { name: /edit/i });
    const unlinkButton = screen.getByRole("button", { name: /unlink/i });
    const deleteButton = screen.getByRole("button", { name: /delete/i });

    expect(editButton).toBeInTheDocument();
    expect(unlinkButton).toBeInTheDocument();
    expect(deleteButton).toBeInTheDocument();

    // Click the buttons and verify callbacks
    await user.click(editButton);
    expect(mockOnEdit).toHaveBeenCalledTimes(1);

    await user.click(unlinkButton);
    expect(mockOnUnlink).toHaveBeenCalledTimes(1);

    await user.click(deleteButton);
    expect(mockOnDelete).toHaveBeenCalledTimes(1);
  });

  test("EquipmentListItem should not render admin buttons when adminView is false", () => {
    render(<EquipmentListItem item={mockEquipment} adminView={false} />);

    // Admin buttons should not be present
    expect(screen.queryByRole("button", { name: /edit/i })).not.toBeInTheDocument();
    expect(screen.queryByRole("button", { name: /unlink/i })).not.toBeInTheDocument();
    expect(screen.queryByRole("button", { name: /delete/i })).not.toBeInTheDocument();
  });

  test("EquipmentListItem should render correct icon based on equipment category", () => {
    const categories = [
      { category: EquipmentCategory.Cellphone, testId: "phone-icon" },
      { category: EquipmentCategory.Tablet, testId: "tablet-icon" },
      { category: EquipmentCategory.Laptop, testId: "laptop-icon" },
      { category: EquipmentCategory.Monitor, testId: "monitor-icon" },
      { category: EquipmentCategory.Headset, testId: "headset-icon" },
      { category: EquipmentCategory.Keyboard, testId: "keyboard-icon" },
    ];

    categories.forEach(({ category, testId }) => {
      const equipment = {
        ...mockEquipment,
        equipmentCatId: category,
      };

      render(<EquipmentListItem item={equipment} />);
      expect(screen.getByTestId(testId)).toBeInTheDocument();
      cleanup();
    });
  });

  test("EquipmentListItem should render unknown device icon for invalid category", () => {
    const equipment = {
      ...mockEquipment,
      equipmentCatId: 999, // Invalid category
    };

    render(<EquipmentListItem item={equipment} />);
    expect(screen.getByTestId("unknown-device-icon")).toBeInTheDocument();
  });
});
