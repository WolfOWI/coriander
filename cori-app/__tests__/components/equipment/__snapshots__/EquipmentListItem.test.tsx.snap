// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots of EquipmentListItem component EquipmentListItem snapshot with admin view 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-2 w-full justify-between group"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-uks7c6-MuiSvgIcon-root"
        data-testid="laptop-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2H0v2h24v-2zM4 6h16v10H4z"
        />
      </svg>
      <div
        class="flex flex-col"
      >
        <p
          class="text-zinc-900"
        >
          MacBook Pro
        </p>
        <p
          class="text-zinc-500 text-sm"
        >
          Laptop
        </p>
      </div>
    </div>
    <p
      class="text-zinc-500 text-sm"
    >
      20 Mar 2024
    </p>
    <div
      class="flex items-center gap-2"
    >
      <div
        class="bg-cyan-600 py-1 px-3 w-fit undefined flex items-center justify-center rounded-lg"
      >
        <span
          class="text-sm text-warmstone-50"
        >
          Good
        </span>
      </div>
      <button
        aria-label="edit"
        class="bg-zinc-500 text-zinc-50 border-2 border-zinc-500 hover:bg-zinc-900 hover:text-zinc-50 hover:border-zinc-900 flex items-center justify-center h-8 w-8 rounded-full cori-btn hidden group-hover:flex transition-all duration-300 btn btn-primary"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="EditIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75z"
          />
        </svg>
      </button>
      <button
        aria-label="unlink"
        class="bg-transparent text-red-500 border-2 border-red-500 hover:border-red-700 hover:text-red-700 flex items-center justify-center h-8 w-8 rounded-full cori-btn hidden group-hover:flex transition-all duration-300 btn btn-outline-primary"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="LinkOffIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17 7h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1 0 1.43-.98 2.63-2.31 2.98l1.46 1.46C20.88 15.61 22 13.95 22 12c0-2.76-2.24-5-5-5m-1 4h-2.19l2 2H16zM2 4.27l3.11 3.11C3.29 8.12 2 9.91 2 12c0 2.76 2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1 0-1.59 1.21-2.9 2.76-3.07L8.73 11H8v2h2.73L13 15.27V17h1.73l4.01 4L20 19.74 3.27 3z"
          />
        </svg>
      </button>
      <button
        aria-label="delete"
        class="bg-red-500 text-white border-2 border-red-500 hover:bg-red-700 hover:border-red-700 flex items-center justify-center h-8 w-8 rounded-full cori-btn hidden group-hover:flex transition-all duration-300 btn btn-primary"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-120dh41-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshots of EquipmentListItem component EquipmentListItem snapshot with basic props 1`] = `
<DocumentFragment>
  <div
    class="flex items-center gap-2 w-full justify-between group"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-uks7c6-MuiSvgIcon-root"
        data-testid="laptop-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 18c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2H0v2h24v-2zM4 6h16v10H4z"
        />
      </svg>
      <div
        class="flex flex-col"
      >
        <p
          class="text-zinc-900"
        >
          MacBook Pro
        </p>
        <p
          class="text-zinc-500 text-sm"
        >
          Laptop
        </p>
      </div>
    </div>
    <p
      class="text-zinc-500 text-sm"
    >
      20 Mar 2024
    </p>
    <div
      class="flex items-center gap-2"
    >
      <div
        class="bg-cyan-600 py-1 px-3 w-fit undefined flex items-center justify-center rounded-lg"
      >
        <span
          class="text-sm text-warmstone-50"
        >
          Good
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
